apis/configwatch.go

```diff
@@ -181,3 +205,101 @@ func MonitorTLS(start time.Time, expiry time.Time, conf *config.Config) {
 		}
 	}
 }
+
+func (s *Server) getClientValidator(helloInfo *tls.ClientHelloInfo) func([][]byte, [][]*x509.Certificate) error {
+
+	log.Debug().Msgf("Inside get client validator - %v", helloInfo.Conn.RemoteAddr())
+	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		// Verifying client certs with root ca
+		opts := x509.VerifyOptions{
+			Roots:         s.CaCertPool,
+			CurrentTime:   time.Now(),
+			Intermediates: x509.NewCertPool(),
+			KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+		}
+		_, err := verifiedChains[0][0].Verify(opts)
+		if err != nil {
+			log.Error().Err(err).Msgf("certs verification failed")
+			return err
+		}
+
+		log.Debug().Msgf("validating with SAN match rules - %s", s.SANMatchRules)
+		if len(s.SANMatchRules) == 0 {
+			log.Info().Msgf("mtls san match rules are undefined")
+			return nil
+		}
+		for _, match := range verifiedChains[0][0].DNSNames {
+			for _, candidateName := range s.SANMatchRules {
```

Comment:
I think we should lower case here for the dns names coming from the SAN and if there is possibility for invalid name according to RFC we should catch here and skip.  Proceed then with the matching below if this check passes.